config/boards/nanopi-r6s.conf

```diff
@@ -40,17 +40,80 @@ function post_family_config__nanopi_r6s_use_mainline_uboot() {
 
 	declare -g BOOTDELAY=1                                       # Wait for UART interrupt to enter UMS/RockUSB mode etc
 	declare -g BOOTSOURCE="https://github.com/u-boot/u-boot.git" # We ❤️ Mainline U-Boot
-	declare -g BOOTBRANCH="tag:v2024.10"
-	declare -g BOOTPATCHDIR="v2024.10"
-	# Don't set BOOTDIR, allow shared U-Boot source directory for disk space efficiency
-
+	declare -g BOOTBRANCH="tag:v2025.01"
+	declare -g BOOTPATCHDIR="v2025.01"
+	declare -g BOOTDIR="u-boot-${BOARD}" # do not share u-boot directory
 	declare -g UBOOT_TARGET_MAP="BL31=${RKBIN_DIR}/${BL31_BLOB} ROCKCHIP_TPL=${RKBIN_DIR}/${DDR_BLOB};;u-boot-rockchip.bin"
-
-	# Disable stuff from rockchip64_common; we're using binman here which does all the work already
-	unset uboot_custom_postprocess write_uboot_platform write_uboot_platform_mtd
+	unset uboot_custom_postprocess write_uboot_platform write_uboot_platform_mtd # disable stuff from rockchip64_common; we're using binman here which does all the work already
 
 	# Just use the binman-provided u-boot-rockchip.bin, which is ready-to-go
 	function write_uboot_platform() {
 		dd "if=$1/u-boot-rockchip.bin" "of=$2" bs=32k seek=1 conv=notrunc status=none
 	}
 }
+
+function pre_config_uboot_target__r6s_patch_uboot_dtsi_for_ums() {
+	display_alert "u-boot for ${BOARD}" "u-boot: add to u-boot dtsi for UMS" "info" # avoid a patch, just append to the dtsi file
+	cat <<- EOD >> arch/arm/dts/rk3588s-nanopi-r6s-u-boot.dtsi
+		&u2phy0 { status = "okay"; };
+		&u2phy0_otg { status = "okay"; };
+		&usbdp_phy0 { status = "okay"; };
+		&usb_host0_xhci { dr_mode = "peripheral";  maximum-speed = "high-speed";  status = "okay"; };
+	EOD
+}
+
+# The upstream DTs (kernel 6.13 / u-boot 2025.01) are in flux and different
+# let's just patch to normalize the SD and eMMC order as in mainline Linux DT: https://github.com/torvalds/linux/blob/master/arch/arm64/boot/dts/rockchip/rk3588s-nanopi-r6.dtsi#L14-L15
+function pre_config_uboot_target__r6s_patch_uboot_dtsi_for_sd_emmc_order() {
+	display_alert "u-boot for ${BOARD}" "u-boot: add to u-boot dtsi for SD=mmc0 and eMMC=mmc1" "info"
+	cat <<- EOD >> arch/arm/dts/rk3588s-nanopi-r6s-u-boot.dtsi
+		/ { aliases { mmc0 = &sdmmc; mmc1 = &sdhci; }; };
+	EOD
+}
+
+# "rockchip-common: boot SD card first, then NVMe, then mmc"
+# include/configs/rockchip-common.h
+# -#define BOOT_TARGETS "mmc1 mmc0 nvme scsi usb pxe dhcp spi"
+# +#define BOOT_TARGETS "mmc0 nvme mmc1 scsi usb pxe dhcp spi"
+# On R6S, mmc1 is the eMMC, mmc0 is the SD card slot
```

Comment:
Maybe USB should have higher priority as well

Reply:
Yeah, in the past I've been bumping USB higher up the list, as it is convenient for people who are experimenting (eg flash Armbian and mainline uboot to eMMC, then use a Fedora on USB, etc). 

But many have complained that `usb start` (enumerating all buses) is quite slow and delays normal boot by quite a few seconds.

For the specific case of the R6*, those don't have too many USB ports, and I'm switching the USB3 port to UMS/Gadget/OTG mode (for UMS/RockUSB/etc) so bumping USB up the list to allow booting from USB2 is not worth it, I think.

(When I do the refactor I intend to make it easy to change the order, and enable/disable OTG so it's easy to make a fast USB3-bootable bootloader).

Reply:
PS. I'm gonna do at least one more rk356x board (`odroid-m1`) and one rk3399 board (`rockpro64`) before any refactoring so I can get the full picture of the complexity of making all those things "generic".